librz/analysis/var_global.c

```diff
@@ -0,0 +1,335 @@
+	return globals;
+}
+
+/**
+ * \brief Rename the global variable
+ * 
+ * \param analysis RzAnalysis
+ * \param old_name The old name of the global variable
+ * \param newname The new name of the global variable
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_rename(RzAnalysis *analysis, RZ_NONNULL const char *old_name, RZ_NONNULL const char *newname) {
+	rz_return_val_if_fail(analysis && old_name && newname, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byname(analysis, old_name);
+	if (!glob) {
+		return false;
+	}
+	RZ_FREE(glob->name);
+	glob->name = strdup(newname);
+	return ht_pp_update_key(analysis->ht_global_var, old_name, newname);
+}
+
+/**
+ * \brief Set the type of the global variable
+ * 
+ * \param glob Global variable instance
+ * \param type The type to set. RzType*
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_set_type(RzAnalysisVarGlobal *glob, RZ_NONNULL RZ_BORROW RzType *type) {
+	rz_return_if_fail(glob && type);
+	glob->type = type;
+}
+
+/**
+ * \brief Add a constaint to global variable
+ * 
+ * \param glob Global variable instance
+ * \param constraint RzTypeConstraint
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_add_constraint(RzAnalysisVarGlobal *glob, RzTypeConstraint *constraint) {
+	rz_return_if_fail(glob && constraint);
+	rz_vector_push(&glob->constraints, constraint);
+}
+
+/**
+ * \brief Get the pritable string of global variable constraints
+ * 
+ * \param glob Global variable instance
+ * \return char *
+ */
+RZ_API RZ_OWN char *rz_analysis_var_global_get_constraints_readable(RzAnalysisVarGlobal *glob) {
+	size_t n = glob->constraints.len;
+	if (!n) {
+		return NULL;
+	}
+	bool low = false, high = false;
+	RzStrBuf sb;
+	rz_strbuf_init(&sb);
+	size_t i;
+	for (i = 0; i < n; i += 1) {
+		RzTypeConstraint *constr = rz_vector_index_ptr(&glob->constraints, i);
+		switch (constr->cond) {
+		case RZ_TYPE_COND_LE:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "<= 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_LT:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "< 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_GE:
+			rz_strbuf_appendf(&sb, ">= 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		case RZ_TYPE_COND_GT:
+			rz_strbuf_appendf(&sb, "> 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		default:
+			break;
+		}
+		if (low && high && i != n - 1) {
+			rz_strbuf_append(&sb, " || ");
+			low = false;
+			high = false;
+		}
+	}
+	return rz_strbuf_drain_nofree(&sb);
+}
+
+/**
+ * \brief Print out the global variables
+ * 
+ * \param analysis RzAnalysis
+ * \param state RzCmdStateOutput *
+ * \param name Which one to print
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_list_show(RzAnalysis *analysis, RzCmdStateOutput *state, RZ_NULLABLE const char *name) {
```

Comment:
I think this whole function should just be moved to canalysis.c. No reason for not doing it here and now I think. It should just be a simple move and you can directly use rz_cons_printf instead of callbacks.